spring-integration-core/src/main/java/org/springframework/integration/router/AbstractMappingMessageRouter.java

```diff
@@ -220,6 +232,9 @@ private void addChannelFromString(Collection<MessageChannel> channels, String ch
 		MessageChannel channel = resolveChannelForName(channelName, message);
 		if (channel != null) {
 			channels.add(channel);
+			if (!mapped && !this.dynamicChannels.contains(channelName)) {
+				this.dynamicChannels.add(channelName);
```

Comment:
I wonder if this logic should be kinda cache with some simple eviction policy in face of expiration time.
The general point that we may route to some channel only once, but it stands in the `List` forever.